Paper: "Phlower: A Deep Learning Framework Supporting PyTorch Tensors with Physical Dimensions"

riku-sakamoto opened the pull request:
This is a pull request to submit our paper titled **"Phlower: A Deep Learning Framework Supporting PyTorch Tensors with Physical Dimensions"** for the SciPy 2025 Proceedings.

The manuscript is now ready for review. I would appreciate it if you could take a look and provide any feedback or suggestions for improvement.

Thank you very much in advance!

riku-sakamoto committed:
[add] add initial paper for phlower

riku-sakamoto committed:
Merge branch 'scipy-conference:2025' into 2025_add_phlower_proceedings

riku-sakamoto committed:
add roles. fix introduction. add DOI

riku-sakamoto committed:
improve

riku-sakamoto committed:
update content

riku-sakamoto committed:
improve codes and links

riku-sakamoto committed:
revise document

riku-sakamoto committed:
minor fix

riku-sakamoto committed:
revise the document

ameyxd commented:
Inviting reviewers: @janakc@amazon.com and @wd275@cornell.edu

ameyxd commented:
Hi folk, I will serve as the editor for this paper when reviews are complete.

Editor: @ameyxd

riku-sakamoto committed:
minor fix

riku-sakamoto commented:
@ameyxd  It seems that no reviewers have been assigned to this PR yet. Could you kindly check if it's proceeding through correct process? 
Please let me know if there's anything I should handle!

ameyxd commented:
Assigned new reviewers, yet to respond - Sanjay Belaturu Krishnegowda, Andy Terrel. In process to reassign reviewers.

ameyxd commented on papers/riku_sakamoto/main.md:
```diff
@@ -0,0 +1,300 @@
+
+As shown in [](#physical_dimension), Physical dimensions are the fundamental quantities used to describe physical systems. These include time ( {math}`T` ), mass ( {math}`M` ), length ( {math}`L` ), electric current ( {math}`I` ), thermodynamic temperature ( {math}`\Theta` ), amount of substance ( {math}`N` ), and luminous intensity ( {math}`J` ). In numerical simulations of physical phenomena, p ...
+
+PyTorch [@NEURIPS2019_bdbca288] is a deep learning framework that provides tensor operations and automatic differentiation capabilities. However, it does not natively support physical dimensions. To fill this gap, we develop Phlower, which is an open-source deep learning framework, and introduce `PhlowerTensor` which is a wrapper class of PyTorch tensors and enables physical dimension tracking. T ...
```
> Thanks to its lightweight implementation, converting between PyTorch's tensor object and `PhlowerTensor` is straightforward.

Might be helpful to mention what in the implementation makes this conversion straightforward in this sentence.

riku-sakamoto replied:
Thanks for the suggestion. I've added some description to clarify this point:

```
From an implementation perspective, `PhlowerTensor` holds an underlying PyTorch tensor, making conversion between PyTorch tensors and `PhlowerTensor` straightforward. 
```

ameyxd commented on papers/riku_sakamoto/main.md:
```diff
@@ -0,0 +1,300 @@
+
+### Enforcing Dimensional Validity in Feature Engineering
+
+Phlower can reject invalid operations in feature engineering (e.g., subtracting pressure from velocity) due to its enforcement of dimensional consistency. This capability is particularly useful in domains where data includes a variety of physical quantities with different dimensions.
```
Can you provide an example of a domain here that this might be applicable for?

riku-sakamoto replied:
I've added an example of domain!

```
This capability is particularly useful in domains such as fluid dynamics (e.g., aerodynamic analysis of vehicles),
```

ameyxd commented on papers/riku_sakamoto/main.md:
```diff
@@ -0,0 +1,300 @@
+
+* **Python Quantities** [@python-quantities]: A library that extends NumPy to support physical quantities with units. It enables dimensional analysis and unit conversions in scientific computations. `Python Quantities` is designed to be compatible with NumPy arrays, allowing seamless integration with NumPy operations.
+
+In contrast to these libraries, Phlower does not aim to provide unit conversion functionalities. Instead, it focuses on enforcing **dimensional consistency** in tensor operations. `PhlowerTensor` is designed to integrate seamlessly with PyTorch, allowing users to leverage existing PyTorch functionalities while tracking and validating the physical dimensions of tensor computations.
```
This is a great point that highlights the utility of Phlower. I would like to see a little more technical detail in this paragraph to contrast with the other libraries mentioned.

riku-sakamoto replied:
It totally makes sense! I've added some technical details of Phlower comparing to other libraries.

```
Second, `PhlowerTensor` represents physical dimensions as an exponent vector of physical quantities (such as mass, length, and time) and stores this metadata alongside a PyTorch tensor. This design enables seamless integration with PyTorch, supporting GPU acceleration and automatic differentiation while tracking and validating the physical dimensions of tensor computations.
```

ameyxd reviewed (commented):
Generally well-written work. I recommend some slight changes to effectively showcase the utility of the work. Good job!

riku-sakamoto committed:
enh: reflect reviewer's comment

riku-sakamoto commented:
@ameyxd 
Thank you for the review! I've updated files to address your comments.

sanjaybk7 commented:
This is a well-written and interesting paper. The motivation is clear, the code examples are effective and the use cases (PINNs, feature engineering) nicely illustrate the value of enforcing physical dimensions in PyTorch workflows.

Only a few minor points to address before final acceptance:

Improve the resolution/legibility of Figures 1 and 2.
Clarify how in-place operations (e.g., tensor += value) behave under dimension checks.
Consider adding a brief note or citation in the Related Work section to further distinguish Phlower from Pint/Quantities.

With these small revisions, I believe the paper is ready to be accepted.

Recommendation: Approve with minor edits

riku-sakamoto committed:
[update] reflect reviewer's suggestions

riku-sakamoto commented:
@sanjaybk7 

Thank you for the thoughtful review! I've updated files to reflect your suggestions.

> Improve the resolution/legibility of Figures 1 and 2.

I have replaced these figures with higher-resolution versions to improve legibility.

> Clarify how in-place operations (e.g., tensor += value) behave under dimension checks.

I've added a description and an example for in-place operation.

> Consider adding a brief note or citation in the Related Work section to further distinguish Phlower from Pint/Quantities.

I've added the following note:

```
Thus, while Pint and Python Quantities aim to provide general-purpose unit handling and conversions in scientific computing built on NumPy, Phlower is specifically designed to integrate dimension checking into deep learning workflows built on PyTorch.
```

Evelyyyynnnn commented:
1.	Novel Contribution
The paper introduces PhlowerTensor, a lightweight wrapper around PyTorch tensors that integrates physical dimension tracking. This is a meaningful contribution to the field of scientific machine learning, where dimensional consistency is essential but often overlooked.
2.	Practical Relevance
By enforcing dimensional consistency at runtime, the framework directly addresses a common source of implementation errors in physics-informed neural networks (PINNs), computational fluid dynamics, and engineering workflows. The integration with PyTorch and compatibility with autograd ensure practical usability.
3.	Well-Structured Presentation
The manuscript is clearly organized, moving logically from motivation and design philosophy to use cases, related work, limitations, and future work. The code examples (Programs 1–8) and YAML/model visualization support reproducibility and user adoption.

Weaknesses and Areas for Improvement
1.	Performance Evaluation
While the runtime overhead of enforcing dimensional consistency is acknowledged, the paper lacks quantitative benchmarks. Profiling Phlower against pure PyTorch workflows (both CPU and GPU) would help readers understand trade-offs in real workloads.
2.	Limited Scope of Demonstrations
The examples focus primarily on toy cases (e.g., kinetic energy, stacking tensors). Demonstrating Phlower within a full physics-informed model (e.g., Navier–Stokes PINN training) would make the framework’s benefits more compelling.
3.	Unit Handling and Conversions
Unlike Pint, Phlower does not support unit conversion. Although this design choice is justified, it may limit adoption in interdisciplinary settings where researchers frequently switch between unit systems. A roadmap for optional unit support would strengthen the paper.

riku-sakamoto commented:
@Evelyyyynnnn  

Thank you for the feedback. Since the editing window has already closed, I won’t be able to incorporate these suggestions into the current version, but I’ll keep them in mind for future work.

ameyxd commented:
> 1. Novel Contribution
>    The paper introduces PhlowerTensor, a lightweight wrapper around PyTorch tensors that integrates physical dimension tracking. This is a meaningful contribution to the field of scientific machine learning, where dimensional consistency is essential but often overlooked.
> 2. Practical Relevance
>    By enforcing dimensional consistency at runtime, the framework directly addresses a common source of implementation errors in physics-informed neural networks (PINNs), computational fluid dynamics, and engineering workflows. The integration with PyTorch and compatibility with autograd ensure practical usability.
> 3. Well-Structured Presentation
>    The manuscript is clearly organized, moving logically from motivation and design philosophy to use cases, related work, limitations, and future work. The code examples (Programs 1–8) and YAML/model visualization support reproducibility and user adoption.
> 
> Weaknesses and Areas for Improvement
> 
> 1. Performance Evaluation
>    While the runtime overhead of enforcing dimensional consistency is acknowledged, the paper lacks quantitative benchmarks. Profiling Phlower against pure PyTorch workflows (both CPU and GPU) would help readers understand trade-offs in real workloads.
> 2. Limited Scope of Demonstrations
>    The examples focus primarily on toy cases (e.g., kinetic energy, stacking tensors). Demonstrating Phlower within a full physics-informed model (e.g., Navier–Stokes PINN training) would make the framework’s benefits more compelling.
> 3. Unit Handling and Conversions
>    Unlike Pint, Phlower does not support unit conversion. Although this design choice is justified, it may limit adoption in interdisciplinary settings where researchers frequently switch between unit systems. A roadmap for optional unit support would strengthen the paper.

This comment does not provide particular edit requests and is vague for the author to make updates on, so will be dismissed.

ameyxd commented:
@riku-sakamoto - This version of the work looks good to me. I appreciate your effort. Approved.

riku-sakamoto commented:
@ameyxd I really appreciate your review and approval. Thank you for your time and support!
